File: Tutorials/WPF/Classic/CS/Wrappers/CustomerEditWrapper.cs
using System;
using DevExpress.Xpo;
using System.ComponentModel;
using XpoTutorial;
using System.Threading.Tasks;
using System.Collections.ObjectModel;
using System.Collections.Generic;

namespace WpfApplication.Wrappers {
    public class CustomerEditWrapper : INotifyPropertyChanged {
        UnitOfWork unitOfWork;
        readonly int? customerOid;

        public CustomerEditWrapper() {
            unitOfWork = new UnitOfWork(XpoDefault.DataLayer);
            Customer = new Customer(unitOfWork);
        }

        public CustomerEditWrapper(int customerOid) {
            this.customerOid = customerOid;
            unitOfWork = new UnitOfWork(XpoDefault.DataLayer);
            Customer = unitOfWork.GetObjectByKey<Customer>(customerOid);
        }

        Customer fCustomer;
        public Customer Customer {
            get {
                return fCustomer;
            }
            set {
                fCustomer = value;
                OrderList = fCustomer.Orders;
                OnPropertyChanged(nameof(Customer));
            }
        }

        IList<Order> fOrderList;
        public IList<Order> OrderList {
            get {
                return fOrderList;
            }
            set {
                fOrderList = value;
                OnPropertyChanged(nameof(OrderList));
            }
        }

        Order fSelectedOrder;
        public Order SelectedOrder {
            get {
                return fSelectedOrder;
            }
            set {
                fSelectedOrder = value;
                IsOrderSelected = (value != null);
                OnPropertyChanged(nameof(SelectedOrder));
            }
        }

        bool fIsOrderSelected;
        public bool IsOrderSelected {
            get {
                return fIsOrderSelected;
            }
            set {
                fIsOrderSelected = value;
                OnPropertyChanged(nameof(IsOrderSelected));
            }
        }

        public Order CreateNewOrder() {
            Order order = new Order(unitOfWork);
            order.Customer = Customer;
            SelectedOrder = order;
            return order;
        }

        public void DeleteSelectedOrder() {
            if(SelectedOrder != null) {
                unitOfWork.Delete(SelectedOrder);
            }
        }

        public async Task ReloadAsync() {
            if(this.customerOid.HasValue) {
                unitOfWork = new UnitOfWork(XpoDefault.DataLayer);
                Customer = await unitOfWork.GetObjectByKeyAsync<Customer>(customerOid);
            }
        }

        public Task SaveAsync() {
            return unitOfWork.CommitChangesAsync();
        }

        public event PropertyChangedEventHandler PropertyChanged;
        void OnPropertyChanged(string propertyName) {
            if(PropertyChanged != null) {
                PropertyChanged(this, new PropertyChangedEventArgs(propertyName));
            }
        }
    }
}

File: Tutorials/WPF/Classic/VB/Wrappers/CustomerEditWrapper.vb
Imports System
Imports DevExpress.Xpo
Imports System.ComponentModel
Imports XpoTutorial
Imports System.Threading.Tasks
Imports System.Collections.ObjectModel
Imports System.Collections.Generic

Namespace WpfApplication.Wrappers
	Friend Class CustomerEditWrapper
		Implements INotifyPropertyChanged

		Private unitOfWork As UnitOfWork
		Private ReadOnly customerOid? As Integer

		Public Sub New()
			unitOfWork = New UnitOfWork(XpoDefault.DataLayer)
			Customer = New Customer(unitOfWork)
		End Sub

		Public Sub New(ByVal customerOid As Integer)
			Me.customerOid = customerOid
			unitOfWork = New UnitOfWork(XpoDefault.DataLayer)
			Customer = unitOfWork.GetObjectByKey(Of Customer)(customerOid)
		End Sub

		Private fCustomer As Customer
		Public Property Customer() As Customer
			Get
				Return fCustomer
			End Get
			Set(ByVal value As Customer)
				fCustomer = value
				OrderList = fCustomer.Orders
				OnPropertyChanged(NameOf(Customer))
			End Set
		End Property

		Private fOrderList As IList(Of Order)
		Public Property OrderList() As IList(Of Order)
			Get
				Return fOrderList
			End Get
			Set(ByVal value As IList(Of Order))
				fOrderList = value
				OnPropertyChanged(NameOf(OrderList))
			End Set
		End Property

		Private fSelectedOrder As Order
		Public Property SelectedOrder() As Order
			Get
				Return fSelectedOrder
			End Get
			Set(ByVal value As Order)
				fSelectedOrder = value
				IsOrderSelected = (value IsNot Nothing)
				OnPropertyChanged(NameOf(SelectedOrder))
			End Set
		End Property

		Private fIsOrderSelected As Boolean
		Public Property IsOrderSelected() As Boolean
			Get
				Return fIsOrderSelected
			End Get
			Set(ByVal value As Boolean)
				fIsOrderSelected = value
				OnPropertyChanged(NameOf(IsOrderSelected))
			End Set
		End Property

		Public Function CreateNewOrder() As Order
			Dim order As New Order(unitOfWork)
			order.Customer = Customer
			SelectedOrder = order
			Return order
		End Function

		Public Sub DeleteSelectedOrder()
			If SelectedOrder IsNot Nothing Then
				unitOfWork.Delete(selectedOrder_Renamed)
			End If
		End Sub

		Public Async Function ReloadAsync() As Task
			If Me.customerOid.HasValue Then
				unitOfWork = New UnitOfWork(XpoDefault.DataLayer)
				Customer = Await unitOfWork.GetObjectByKeyAsync(Of Customer)(customerOid)
			End If
		End Function

		Public Function SaveAsync() As Task
			Return unitOfWork.CommitChangesAsync()
		End Function

		Public Event PropertyChanged As PropertyChangedEventHandler Implements INotifyPropertyChanged.PropertyChanged
		Private Sub OnPropertyChanged(ByVal propertyName As String)
			RaiseEvent PropertyChanged(Me, New PropertyChangedEventArgs(propertyName))
		End Sub
	End Class
End Namespace